exp/topology/desiredstate/lifecycle_hooks.go

```diff
@@ -96,6 +97,45 @@ func (g *generator) callBeforeClusterUpgradeHook(ctx context.Context, s *scope.S
 	return true, nil
 }
 
+// callBeforeControlPlaneUpgradeHook calls the BeforeControlPlaneUpgrade before the control plane picks up a new control plane version,
+// no matter if this is an intermediate versions of an upgrade plan or the target version of an upgrade plan.
+// NOTE: when an upgrade starts, the hook should be called after the BeforeClusterUpgrade hook.
+// NOTE: the hook doesn't need call intent tracking: it is always called before picking up a new control plane version.
+func (g *generator) callBeforeControlPlaneUpgradeHook(ctx context.Context, s *scope.Scope, currentVersion *string, nextVersion string) (bool, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// NOTE: the hook should always be called before piking up a new version.
+	v1beta1Cluster := &clusterv1beta1.Cluster{}
+	// DeepCopy cluster because ConvertFrom has side effects like adding the conversion annotation.
+	if err := v1beta1Cluster.ConvertFrom(s.Current.Cluster.DeepCopy()); err != nil {
+		return false, errors.Wrap(err, "error converting Cluster to v1beta1 Cluster")
+	}
+
+	hookRequest := &runtimehooksv1.BeforeControlPlaneUpgradeRequest{
+		Cluster:               *cleanupCluster(v1beta1Cluster),
+		FromKubernetesVersion: *currentVersion,
+		ToKubernetesVersion:   nextVersion,
+		ControlPlaneUpgrades:  toUpgradeStep(s.UpgradeTracker.ControlPlane.UpgradePlan),
+		WorkersUpgrades:       toUpgradeStep(s.UpgradeTracker.MachineDeployments.UpgradePlan, s.UpgradeTracker.MachinePools.UpgradePlan),
+	}
+	hookResponse := &runtimehooksv1.BeforeControlPlaneUpgradeResponse{}
+	if err := g.RuntimeClient.CallAllExtensions(ctx, runtimehooksv1.BeforeControlPlaneUpgrade, s.Current.Cluster, hookRequest, hookResponse); err != nil {
+		return false, err
+	}
+	// Add the response to the tracker so we can later update condition or requeue when required.
```

Comment:
I extended the condition sub-task on the umbrella issue that we actually do that for the new hooks.
I think the code there should already do mostly the right thing. But we should still take a look, maybe improve messages and add test coverage

Reply:
💯 
thanks for updating the umbrella issue